app/assets/javascripts/discourse/app/components/time-shortcut-picker.js

```diff
@@ -0,0 +1,250 @@
+import { action } from "@ember/object";
+import { and } from "@ember/object/computed";
+
+// global shortcuts that interfere with these modal shortcuts, they are rebound when the
+// component is destroyed
+//
+// c createTopic
+// r replyToPost
+// l toggle like
+// t replyAsNewTopic
+const GLOBAL_SHORTCUTS_TO_PAUSE = ["c", "r", "l", "t"];
+const BINDINGS = {
+  "l t": {
+    handler: "selectShortcut",
+    args: [TIME_SHORTCUT_TYPES.LATER_TODAY],
+  },
+  "l w": {
+    handler: "selectShortcut",
+    args: [TIME_SHORTCUT_TYPES.LATER_THIS_WEEK],
+  },
+  "n d": {
+    handler: "selectShortcut",
+    args: [TIME_SHORTCUT_TYPES.TOMORROW],
+  },
+  "n w": {
+    handler: "selectShortcut",
+    args: [TIME_SHORTCUT_TYPES.NEXT_WEEK],
+  },
+  "n b w": {
+    handler: "selectShortcut",
+    args: [TIME_SHORTCUT_TYPES.START_OF_NEXT_BUSINESS_WEEK],
+  },
+  "n m": {
+    handler: "selectShortcut",
+    args: [TIME_SHORTCUT_TYPES.NEXT_MONTH],
+  },
+  "c r": { handler: "selectShortcut", args: [TIME_SHORTCUT_TYPES.CUSTOM] },
+  "n r": { handler: "selectShortcut", args: [TIME_SHORTCUT_TYPES.NONE] },
+};
+
+export default Component.extend({
+  tagName: "",
+
+  userTimezone: null,
+
+  onTimeSelected: null,
+
+  selectedShortcut: null,
+  selectedTime: null,
+  selectedDate: null,
+  selectedDatetime: null,
+  prefilledDatetime: null,
+
+  additionalOptionsToShow: [],
+  hiddenOptions: [],
+  customOptions: [],
+
+  lastCustomDate: null,
+  lastCustomTime: null,
+  parsedLastCustomDatetime: null,
+  customDate: null,
+  customTime: null,
+
+  defaultCustomReminderTime: `0${START_OF_DAY_HOUR}:00`,
+
+  @on("init")
+  _setupPicker() {
+    this.setProperties({
+      customTime: this.defaultCustomReminderTime,
+      userTimezone: this.currentUser.resolvedTimezone(this.currentUser),
+    });
+
+    if (this.prefilledDatetime) {
+      let parsedDatetime = parseCustomDatetime(
+        this.prefilledDatetime,
+        null,
+        this.userTimezone
+      );
+
+      if (parsedDatetime.isSame(laterToday())) {
+        return this.set("selectedShortcut", TIME_SHORTCUT_TYPES.LATER_TODAY);
+      }
+
+      this.setProperties({
+        customDate: parsedDatetime.format("YYYY-MM-DD"),
+        customTime: parsedDatetime.format("HH:mm"),
+        selectedShortcut: TIME_SHORTCUT_TYPES.CUSTOM,
+      });
+    }
+
+    this._bindKeyboardShortcuts();
+    this._loadLastUsedCustomDatetime();
+  },
+
+  @on("willDestroyElement")
+  _resetKeyboardShortcuts() {
+    KeyboardShortcuts.unbind(BINDINGS);
+    KeyboardShortcuts.unpause(GLOBAL_SHORTCUTS_TO_PAUSE);
+  },
+
+  _loadLastUsedCustomDatetime() {
+    let lastTime = localStorage.lastCustomTime;
+    let lastDate = localStorage.lastCustomDate;
+
+    if (lastTime && lastDate) {
+      let parsed = parseCustomDatetime(lastDate, lastTime, this.userTimezone);
+
+      if (parsed < now(this.userTimezone)) {
+        return;
+      }
+
+      this.setProperties({
```

Comment:
Honestly not a big fan of private function setting properties, that's a smell in my book

Reply:
Can you elaborate on this? Private methods in ruby are used to set instance variables all the time, what is the difference here? Am I misunderstanding what these properties are meant for? What would the alternative be in this case?

Reply:
I would argue it would be good in ruby too but concerning our subject. You are in a component, so properties here are basically impacting UI, my argument is that private functions should more or less be pure functions. Computed properties are not just holding values they can trigger re-rendering of the UI so it's best to make it very clear where it's happening and not all over your code. But up to you, I'm not 100% requesting this to be changed.

Reply:
Right I see that makes sense...I will think about how it could be clearer